kubernetes-client-api/src/test/java/io/fabric8/kubernetes/client/ConfigSourcePrecedenceTest.java

```diff
@@ -0,0 +1,220 @@
+
+        // When + Then
+        assertThat(config)
+            .hasFieldOrPropertyWithValue("masterUrl", "https://user-configuration-override:8443/")
+            .hasFieldOrPropertyWithValue("namespace", "namespace-overridden")
+            .hasFieldOrPropertyWithValue("autoOAuthToken", "token-overridden");
+      } finally {
+        System.clearProperty("kubernetes.master");
+        System.clearProperty("kubernetes.namespace");
+        System.clearProperty("kubernetes.auth.token");
+      }
+    }
+
+    @AfterEach
+    void tearDown() {
+      System.clearProperty("KUBERNETES_SERVICE_HOST");
+      System.clearProperty("KUBERNETES_SERVICE_PORT");
+      System.clearProperty("kubernetes.auth.serviceAccount.token");
+      System.clearProperty("kubenamespace");
+      System.clearProperty("kubeconfig");
+    }
+  }
+
+  @Nested
+  @DisplayName("System Properties")
+  class SystemPropertiesSource {
+    @BeforeEach
+    void setUp() {
+      System.setProperty("kubernetes.master", "https://user-configuration-override:8443");
+      System.setProperty("kubernetes.namespace", "namespace-overridden");
+      System.setProperty("kubernetes.auth.token", "token-overridden");
+    }
+
+    @Test
+    @DisplayName("when no other sources configured, then read from System properties")
+    void whenNoOtherSourceProvided_thenUseSystemProperties() {
+      assertThat(new ConfigBuilder().build())
+          .hasFieldOrPropertyWithValue("masterUrl", "https://user-configuration-override:8443/")
+          .hasFieldOrPropertyWithValue("namespace", "namespace-overridden")
+          .hasFieldOrPropertyWithValue("autoOAuthToken", "token-overridden");
+    }
+
+    @Nested
+    @DisplayName("User configuration overrides Config attributes configured via System Properties")
+    class UserConfigurationOverSystemProperties extends UserConfigurationViaConfigBuilder {
+    }
+
+    @AfterEach
+    void tearDown() {
+      System.clearProperty("kubernetes.master");
+      System.clearProperty("kubernetes.namespace");
+      System.clearProperty("kubernetes.auth.token");
+    }
+  }
+
+  private abstract static class UserConfigurationViaConfigBuilder {
+    @SuppressWarnings("unused")
+    @Test
+    @DisplayName("User configuration via builder given most precedence")
+    void whenUserConfigurationOverridesSomeFields_thenUserConfigurationGivenPrecedence() {
+      // Given
+      Config config = new ConfigBuilder()
+          .withMasterUrl("https://user-configuration-override:8443")
+          .withNamespace("namespace-overridden")
+          .withAutoOAuthToken("token-overridden")
+          .build();
+
+      // When + Then
+      assertThat(config)
+          .hasFieldOrPropertyWithValue("masterUrl", "https://user-configuration-override:8443/")
+          .hasFieldOrPropertyWithValue("namespace", "namespace-overridden")
+          .hasFieldOrPropertyWithValue("autoOAuthToken", "token-overridden");
+    }
+  }
```

Comment:
It's not worth reusing this test.
It obfuscates things and makes it hard to understand what's going on, especially in case of failure.
In addition, there's no check for precedence (as suggested in one of the other comments). You'd specifically want one assertion to ensure that 1. some values are overridden, and 2. some values are preserved from the precedent config